hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java

```diff
@@ -1,167 +0,0 @@
- *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package com.hazelcast.internal.serialization.impl;
-
-import com.hazelcast.nio.ObjectDataInput;
-import com.hazelcast.nio.serialization.DataSerializable;
-import com.hazelcast.nio.serialization.impl.Versioned;
-import com.hazelcast.test.annotation.ParallelJVMTest;
-import com.hazelcast.test.annotation.QuickTest;
-import com.hazelcast.test.starter.GuardianException;
-import com.hazelcast.test.starter.HazelcastVersionLocator;
-import com.hazelcast.internal.util.StringUtil;
-import com.hazelcast.version.Version;
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.experimental.categories.Category;
-import org.junit.rules.TemporaryFolder;
-import org.junit.runner.RunWith;
-import org.mockito.Mockito;
-import org.powermock.core.classloader.annotations.PowerMockIgnore;
-import org.powermock.core.classloader.annotations.PrepareForTest;
-import org.powermock.modules.junit4.PowerMockRunner;
-import org.reflections.Reflections;
-import org.reflections.scanners.SubTypesScanner;
-import org.reflections.util.ConfigurationBuilder;
-
-import java.io.File;
-import java.net.URL;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Set;
-
-import static com.hazelcast.instance.BuildInfoProvider.getBuildInfo;
-import static com.hazelcast.internal.cluster.Versions.CURRENT_CLUSTER_VERSION;
-import static com.hazelcast.internal.cluster.Versions.PREVIOUS_CLUSTER_VERSION;
-import static com.hazelcast.test.HazelcastTestSupport.assumeThatNoJDK6;
-import static com.hazelcast.test.HazelcastTestSupport.assumeThatNoJDK7;
-import static com.hazelcast.test.ReflectionsHelper.REFLECTIONS;
-import static com.hazelcast.test.ReflectionsHelper.filterNonConcreteClasses;
-import static com.hazelcast.test.starter.HazelcastStarterUtils.rethrowGuardianException;
-import static com.hazelcast.internal.util.EmptyStatement.ignore;
-import static org.junit.Assert.fail;
-import static org.junit.Assume.assumeNoException;
-import static org.mockito.Mockito.spy;
-import static org.mockito.Mockito.when;
-
-/**
- * Tests a common compatibility issue: when a (Identified)DataSerializable class first
- * becomes Versioned, it might miss checking input stream for UNKNOWN version (which is
- * the version of an incoming stream from a previous-version member) instead of using
- * in.getVersion.isUnknownOrLessThan(CURRENT).
- */
-@RunWith(PowerMockRunner.class)
-@PowerMockIgnore({"javax.net.ssl.*", "javax.security.*", "javax.management.*"})
-@PrepareForTest(Version.class)
-@Category({QuickTest.class, ParallelJVMTest.class})
-public class VersionedInCurrentVersionTest {
```

Comment:
@vbekiaris could you confirm my understanding that this test is no longer needed

Reply:
@ldziedziul can you explain why do you think this test is no longer needed?

Reply:
Here's the dump from my conversation with @vbekiaris: 

The reason I'm asking is that in the current form it's impossible to make this test work with JDK17 without opening `java.io` module for tests (which brings some danger of using `java.io` internals in production code).

I've noticed `in.getVersion()` is called only with subsequent `isGreaterOrEqual` and `isGreaterThan` calls

`Version#isUnknownOrLessThan` is called only in
* `RaftService#publishGroupAvailabilityEvents`
* `AbstractDistributedObject#isClusterVersionUnknownOrLessThan`

On the other hand `Version#isLessThan` is used only here:
* `ClusterWideConfigurationService#checkConfigVersion`
* `AbstractDistributedObject#isClusterVersionLessThan`
* `SqlServiceImpl#execute`

I see no places where this test could fail.

From @vbekiaris: 
> The purpose of the test is to avoid allowing someone to write new code using `in.getVersion().isLessThan` in an IDS class’ `readData` that was not `Versioned` in the previous minor version and became `Versioned` in the current codebase.
> can we somehow ensure that and not open `java.io`?

From me:
If we would remove `final` from `Version` then regular (non-powermockito) spy would work

I've also came with another solution: an Archunit rule:

```java
   noClasses().should()
                .callMethod(Version.class, "isLessThan", Version.class)
                .because("newly versioned classes should use 'isUnknownOrLessThan' or similar instead")
                .check(versionedInCurrentVersion);
```

it works but only for direct usage of `in.getVersion().isLessThan` in classes from `versionedInCurrentVersion`. If an input stream is passed to another class then call to `isLessThan` won't be detected. I don't know if it's an issue or not.

To sum it up, if we want to keep this test and avoid `--add-opens` in tests then we have two ways:
- remove `final` from `Version`
- test only direct call using ArchUnit
- something else.... ? e.g. serializing an old class first and then trying to deserialize it with a new class?

What do you think?

Reply:
I believe for the given cases `SerializedObjectsCompatibilityTest` should also fail. So I now believe we can remove this test, if keeping isn't required.

Reply:
Good point @ramizdundar , I agree with  you